### DISCUSSION TITLE
Question: keeping client connections alive after input stream interruptions

### DISCUSSION THREAD

**Discussion by derSoerrn95** [bluenviron/gortsplib#791] [Questions]
**Description:**
I'm new to RTSP and streaming. I’ve tried the server example in the `gortsplib` repo (https://github.com/bluenviron/gortsplib/blob/main/examples/server/main.go), but the clients get terminated when the input stream ends. My application already automatically restarts the FFmpeg process that sends the RTSP input stream. 

I think it's because the `sh.stream.Close()` when a new stream is announced what happens, when I'm restarting my ffmpeg input process.

I would like to build a RTSP server that:
- Receives an RTSP stream from an FFmpeg process.
- Allows multiple FFmpeg clients to connect and receive the stream.
- Keeps the FFmpeg clients connected even if the input stream is interrupted (e.g., network failure or FFmpeg process termination).
- Automatically resumes feeding the stream to clients once the input stream is restored without requiring clients to reconnect.

**Example FFmpeg processes:**

1. **Input FFmpeg process (sending to RTSP server):**
   ```bash
   ffmpeg -i input_stream_url_cameraX -c:v copy -c:a copy -f rtsp rtsp://localhost:8554/{{ cameraX }}/stream
   ```
2. **Client FFmpeg process (receiving from RTSP server and outputting to another stream):**
    ```bash
    ffmpeg -i rtsp://localhost:8554/{{ cameraX }}/stream -c:v libx264 -c:a aac -f flv rtmp://cameraX/stream
    ```

**Questions:**
- Is this possible with `gortsplib`?
- How can I maintain the connection to the clients when the ffmpeg process that sends the input stream is restarted?

**Comment by aler9** [comment]
It is possible indeed, it's enough to create `ServerStream` once, delete every `stream.Close()` reference and reuse it.
The problem is that most clients close the connection in case of pauses of incoherent streams, that is what happens when you switch from a publisher to the next one.

The stream can be made more coherent by manually filling `PayloadType`, `SequenceNumber` and `Timestamp` of RTP packets in order to produce an homogeneous stream, for instance:

```go
	nextSequenceNumber := 0

	// called when receiving a RTP packet
	ctx.Session.OnPacketRTPAny(func(medi *description.Media, _ format.Format, pkt *rtp.Packet) {
		// make sure that PayloadType is the same advertised by the first publisher
		pkt.PayloadType = 96

		// recompute SequenceNumber
		pkt.SequenceNumber = nextSequenceNumber++

		// recompute Timestamp
		pkt.Timestamp = (monothonic timestamp with a 90khz clock)

		// route the RTP packet to all readers
		sh.stream.WritePacketRTP(medi, pkt) //nolint:errcheck
	})
```
